### PR TITLE
implemented foldL/R, scanL/R, and reduceL/R. added with tests

### PR DESCRIPTION
**Pull request by ianzapolsky**

### CODE REVIEW

**Comment by ianzapolsky** [comment]
this branch contains all the advanced list operations we talked about. It is important to draw distinctions between fold, reduce, and scan.

compiled from [several](http://oldfashionedsoftware.com/2009/07/10/scala-code-review-foldleft-and-foldright/) [places](http://stackoverflow.com/questions/17408880/reduce-fold-or-scan-left-right) on the internet, this is my understanding:

fold takes an initial value and a function, returning a value that is of the same type as its initial value
reduce takes no initial value, but does take a function, and returns a value that is of the same type as that stored in the list
scan is just like fold; it takes an initial value and a function, but it stores the intermediate values of the fold in a new list and returns that.

these functions are implemented here.

**Comment by ianzapolsky** [comment]
so it seems from the benchmarks that our multithreaded calls are actually performing worse than the linear ones. any idea why this is? I thought that pthreads would be a pretty fast way of doing things because they seem closer to the metal. perhaps I was mistaken

**Comment by natebrennand** [comment]
I have some stuff that improves the performance a bit that I'll PR after we get this in. Basically threads make your code slower unless the workload is significant and software threads == hardware threads.

File: src/collections.h [hunk 1]
```diff
@@ -46,66 +46,122 @@ class Collection {
...
+    foldRight(Function func, U init);
+
+    // Returns the intermediate results of the binary cumulation of the 
+    // elements in a Collection as well as an initial value, starting from the
+    // left
+    template<typename Function, typename U>
+    Collection<typename std::result_of<Function(U, T)>::type>
+    scanLeft(Function func, U init);
+
+    // Returns the intermediate results of the binary cumulation of the
```
Comment (natebrennand) [inline src/collections.h]: s/cumulation/accumulation

File: src/collections.h [hunk 2]
```diff
@@ -46,66 +46,122 @@ class Collection {
...
+
+    // Return the result of the application of the same binary operator on
+    // all elements in the Collection as well as an initial value, starting 
+    // from the right 
     template<typename Function, typename U>
     typename std::result_of<Function(U, T)>::type
-    reduce(Function func, U init);
+    foldRight(Function func, U init);
+
+    // Returns the intermediate results of the binary cumulation of the
```
Comment (natebrennand) [inline src/collections.h]: s/cumulation/accumulation

File: src/collections.h [hunk 3]
```diff
@@ -313,6 +369,83 @@ Collection<T>::fold(std::function<T(T, T)> func) {
...
+        start = end;
+    }
+
+    for (int i = 0; i < threads; i++)
+        pthread_create(&(thread_pool[i]), NULL, preduce_thread<T>, &(thread_data_pool[i]));
+
+    for (pthread_t i : thread_pool)
+        pthread_join(i, NULL);
+
+    T val = func(thread_data_pool[0].retval, thread_data_pool[1].retval);
```
Comment (natebrennand) [inline src/collections.h]: we need bounds checking, here. wanna add a not so we do it later w/ every other method?

File: src/collections.h [hunk 4]
```diff
@@ -356,16 +489,10 @@ Collection<T>::zipWith(Function func, Collection<U>... other_list) {
...
-    static_assert(
-        std::is_same<
-            decltype(func),
-            std::function<U(U, T)>
-        >::value,
-           "Reduce fn must return the same type as the initial value");
-    */
+Collection<T>::foldLeft(Function func, U init) {
+    using return_type = typename std::result_of<Function(U, T)>::type;
+    static_assert(std::is_same<return_type, U>::value,
```
Comment (natebrennand) [inline src/collections.h]: nice

File: src/tests/fail_foldRight.cpp [hunk 5]
```diff
@@ -0,0 +1,14 @@
+#include <vector>
+#include <iostream>
+#include "../collections.h"
+
+
+int main() {
+    Collection<int> ints = Collection<int>(std::vector<int> {2, 3, 4, 5});
+
```
Comment (natebrennand) [inline src/tests/fail_foldRight.cpp]: Can you add a comment detailing why this should fail?
It'll make things easier to maintain

File: src/tests/pass_preduce.cpp [hunk 6]
```diff
@@ -0,0 +1,16 @@
+#include <vector>
+#include <iostream>
+#include <cassert>
+
+
+#include "../collections.h"
+
+
+int main(){
+    auto ints = Collection<int>(std::vector<int> {1, 2, 3, 4, 5, 6, 7, 8, 9, 10});
```
Comment (natebrennand) [inline src/tests/pass_preduce.cpp]: you can just use `range` here!

**Comment by natebrennand** [comment]
:+1: after those comments are resolved

**Comment by natebrennand** [comment]
:+1: pending the build

**Pull request merged**